primitives/src/witness.rs

```diff
@@ -300,36 +300,47 @@ impl Encoder for WitnessEncoder<'_> {
 /// The decoder for the [`Witness`] type.
 #[cfg(feature = "alloc")]
 pub struct WitnessDecoder {
-    /// A decoder for the initial length prefix and subsequent per-element prefixes.
-    prefix_decoder: Option<CompactSizeDecoder>,
-    /// Holds the elements.
-    buffer: Vec<Vec<u8>>,
-    /// True if the initial compact size has been read.
-    initial_length_prefix_read: bool, // I.e not the one for each element.
-    /// Set after the initial length prefix is read.
-    ///
-    /// This is read as a u64, checked to be below 4,000,000 then
-    /// cast to a `usize` to make usage easier.
-    witness_elements: usize,
-    /// Index of the element we are going to decode next.
-    idx: usize,
-    /// True if the element length prefix has been read.
-    element_length_prefix_read: bool,
-    /// Bytes left to read for this element.
-    bytes_to_read: usize,
+    /// The single buffer that will become the Witness content.
+    /// The index entries are written at the beginning, then rotated in [`Self::end`].
+    content: Vec<u8>,
+    /// Current write position in the content buffer.
+    cursor: usize,
+    /// Decoder for the initial witness element count.
+    witness_count_decoder: CompactSizeDecoder,
+    /// Total number of witness elements to decode (None until initial count is read).
+    witness_elements: Option<usize>,
+    /// Index of the current element being decoded.
+    element_idx: usize,
+    /// Decoder for the current element's length.
+    element_length_decoder: CompactSizeDecoder,
+    /// Bytes remaining to read for the current element's data.
+    /// - `None` means we're currently reading the length.
+    /// - `Some(n)` means we're reading element data with `n` bytes remaining.
+    element_bytes_remaining: Option<usize>,
 }
 
 impl WitnessDecoder {
     /// Constructs a new witness decoder.
     pub fn new() -> Self {
         Self {
-            prefix_decoder: None,
-            buffer: Vec::new(),
-            initial_length_prefix_read: false,
-            witness_elements: 0,
-            idx: 0,
-            element_length_prefix_read: false,
-            bytes_to_read: 0,
+            content: Vec::new(),
+            cursor: 0,
+            witness_elements: None,
+            witness_count_decoder: CompactSizeDecoder::new(),
+            element_idx: 0,
+            element_length_decoder: CompactSizeDecoder::new(),
+            element_bytes_remaining: None,
+        }
+    }
+
+    /// Resizes the content buffer if needed, doubling the size each time.
+    fn resize_if_needed(&mut self, required_len: usize) {
```

Comment:
This resize logic is copied from the old decoder, but should maybe delegate to some `consensus_encoding` defined thing like @jrakibi is working on?